### PR TITLE
Add implementation and contract tests for new AWS resources

### PR DESCRIPTION
**Pull request by ezhang6811**
*Description of changes:*
Support instrumentation for the following attributes. See [this Python PR](https://github.com/aws-observability/aws-otel-python-instrumentation/pull/265) for the source of feature parity.

* For SNS, extract `TopicArn` from request body and add to span as `aws.sns.topic.arn`.
* For Secrets Manager, extract either `SecretId` from request body or `ARN` from response body and add to span as `aws.secretsmanager.secret.arn`.
* For Step Functions, extract `activityArn` or  `stateMachineArn` from request body, and add to span as `aws.stepfunctions.activity.arn` or `aws.stepfunctions.state_machine.arn` respectively.
* For Lambda, extract `UUID` from request body for event source mapping functions, and add to span as `aws.lambda.resource_mapping.id`.

In addition, the new resource attribute `aws.remote.resource.cfn.primary.identifier` is also added for each new resource.

E2E tests show that the new attributes and their associated CFN primary identifiers have been correctly populated for each service.

SNS:
image

Secrets Manager:
image

Step Functions:
image

image

Lambda:
image

Contract tests verify the expected span attribute values with example function calls that involve each resource. https://github.com/aws-observability/aws-otel-dotnet-instrumentation/actions/runs/12127063253/job/33813330563?pr=143

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: test/contract-tests/tests/test/amazon/awssdk/awssdk_test.py [hunk 1]
```diff
@@ -558,12 +794,16 @@ def test_bedrock_agent_get_data_source(self):
...
             remote_resource_identifier="test-data-source",
+            cloudformation_primary_identifier="test-knowledge-base|test-data-source",
             request_response_specific_attributes={
                 _AWS_BEDROCK_DATA_SOURCE_ID: "test-data-source",
+                _AWS_BEDROCK_KNOWLEDGE_BASE_ID: "test-knowledge-base"
             },
             span_name="Bedrock Agent.GetDataSource",
         )
 
+    # TODO: add contract test for Lambda event source mapping resource
```
Comment (ezhang6811) [inline test/contract-tests/tests/test/amazon/awssdk/awssdk_test.py:805]: Leaving a TODO item here for the event source mapping contract test case. When creating an event source mapping, LocalStack generates a random UUID each time which is then needed to call GetEventSourceMapping. Still looking for a solution for passing this UUID into the API call after it is generated.

**Review by mxiamxia — APPROVED** [review]

**Comment by vastin** [comment]
Both this PR and PR #142 add `aws.remote.resource.cfn.primary.identifier`. Is there any dependency between them? Should PR #142 be merged firstly?

**Comment by ezhang6811** [comment]
> Both this PR and PR #142 add `aws.remote.resource.cfn.primary.identifier`. Is there any dependency between them? Should PR #142 be merged firstly?

Yes, #142 should be merged first as it originally implements `aws.remote.resource.cfn.primary.identifier`, and this PR was rebased on #142 so that this CFN Identifier attribute could be applied for the new resources.

File: test/AWS.Distro.OpenTelemetry.AutoInstrumentation.Tests/AwsMetricAttributesGeneratorTest.cs
[the remark this answers is not in the document]
Comment (ezhang6811) [reply test/AWS.Distro.OpenTelemetry.AutoInstrumentation.Tests/AwsMetricAttributesGeneratorTest.cs:1]: These warnings aren't related to the changes in this PR. I can put up a separate PR to address them.

File: src/AWS.Distro.OpenTelemetry.AutoInstrumentation/AwsMetricAttributeGenerator.cs [hunk 2]
```diff
@@ -431,15 +475,19 @@ private static void SetRemoteResourceTypeAndIdentifier(Activity span, ActivityTa
...
-            {
-                remoteResourceType = NormalizedBedrockServiceName + "::KnowledgeBase";
-                remoteResourceIdentifier = EscapeDelimiters((string?)span.GetTagItem(AttributeAWSBedrockKnowledgeBaseId));
-            }
             else if (IsKeyPresent(span, AttributeAWSBedrockDataSourceId))
             {
                 remoteResourceType = NormalizedBedrockServiceName + "::DataSource";
                 remoteResourceIdentifier = EscapeDelimiters((string?)span.GetTagItem(AttributeAWSBedrockDataSourceId));
+                cloudformationPrimaryIdentifier =
+                    EscapeDelimiters((string?)span.GetTagItem(AttributeAWSBedrockKnowledgeBaseId))
```
[the remark this answers is not in the document]
Comment (ezhang6811) [reply src/AWS.Distro.OpenTelemetry.AutoInstrumentation/AwsMetricAttributeGenerator.cs:483]: added CFN primary identifier coverage for all unit tests.

**Review by vastin — APPROVED** [review]

**Pull request merged**